### PR TITLE
Rework dynamic forms topic for style and organization

### PR DESCRIPTION
**Pull request by jbogarthyde**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
https://angular.io/guide/dynamic-form does not conform to current doc standards for style and task orientation.

Issue Number:  https://angular-team.atlassian.net/browse/DOCS-726

## What is the new behavior?
Rework, edit, re-organize content to conform to current doc standards

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: aio/content/guide/dynamic-form.md [hunk 1]
```diff
@@ -1,42 +1,41 @@
...
+* Basic knowledge of [reactive forms](guide/reactive-forms "Reactive forms guide").
 
-{@a bootstrap}
+## Setup
 
-## Bootstrap
+Dynamic forms are based on reactive forms. To give the application access reactive forms directives, the [root module](guide/bootstrapping "Learn about bootstrapping an app from the root module.") imports `ReactiveFormsModule` from the `@angular/forms` library.
 
-Start by creating an `NgModule` called `AppModule`.
+The example used here to illustrate the dynamic-forms technique is a reactive form that you use to build an online application experience for heroes seeking employment.
```
Comment (aikidave) [inline aio/content/guide/dynamic-form.md:28]:
```suggestion
In this tutorial, you will build a form using the dynamic forms technique. You will build an online application for heroes seeking employment.
```

**Comment by aikidave** [comment]
Hi Judy,

I think this topic has a lot of great content in it. As I read it, however, I think it would resonate better with users as a tutorial. Could you restructure it as such? The content is almost written as a tutorial anyway, and this could really help users understand the concept.

Here's my definition of a tutorial and what it should contain (pasted from a doc I have about this--apologies for any formatting issues!):

A tutorial takes a prescriptive approach to explaining a concept or pattern to the intended audience. Tutorials contain the following sections.

**Overview**
The overview section has the following requirements:

- Tell the user in the first sentence what the tutorial is about.
- Tell the user what they're going to learn and provide any concise background information that's helpful.
- Don't use the heading "Overview." Just get right to it.

**Objectives**
Provide a short, bulleted list of what  users will learn from the tutorial.

**Prerequisites**
List the essential prerequisites users need to know. Keep in mind our audience types. Feel free to list an existing tutorial as a prerequisite. For example, a tutorial on routing might list the Tour of Heroes introductory tutorial as its primary, perhaps only, prerequisite.

**Body**

- Use as many headings and subheadings as needed.
- Use numbered steps in each section.
- Keep numbered step lists to around 7 or less, if possible. If you need more steps, break it up into subheadings.
- Provide context and explain what's going on.
- Use screenshots when they help the reader. Don't provide a screenshot for every step.
- Show what success looks like along the way. For example, showing console output or describing what happens helps the reader to feel like they're doing it right and help them know things are working so far.

**Next steps**
Provide links to other topics. Tasks that users complete in a tutorial should likely exist as task-based guides, for example. Concepts described in a tutorial should exist as conceptual topics. If these topics don't exist yet, file issues so we can address them.

**Review by aikithoughts — COMMENTED** [review]
Is this still in progress? Looks like it still needs a "Next steps" section at the end?

**Review by aikithoughts — APPROVED** [review]

File: aio/content/guide/dynamic-form.md [hunk 2]
```diff
@@ -1,42 +1,51 @@
...
-It might evolve to support a much richer variety of questions, more graceful rendering, and superior user experience.
-All such greatness has humble beginnings.
+The tutorial walks you through the following steps.
 
-The example in this cookbook is a dynamic form to build an
-online application experience for heroes seeking employment.
-The agency is constantly tinkering with the application process.
-You can create the forms on the fly *without changing the application code*.
-{@a toc}
+1. Set up the data model behind the questionaire, defining some heroes and their attributes.
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:20]: I noticed that the list of steps here doesn't 1:1 reflect the actual sections in the doc below. While this might be treated as a separate logical unit describing the process (vs actual step-by-step instructions), I think it'd be great to keep them in sync so that it also serves as a kind of table of contents. I understand that the steps outlined here require more content than just a few words (that works for section headers below), but may be we can rephrase them to be more aligned? I was thinking of something like this:

- **Create data models to represent form controls**
Set up the data model behind the questionaire, defining some heroes and their attributes.

- **Compose dynamic form contents**
Compose a set of question components into a dynamic-form container component

- ...

File: aio/content/guide/dynamic-form.md [hunk 3]
```diff
@@ -50,79 +59,58 @@ Bootstrap the `AppModule` in `main.ts`.
...
+You can bind a form to these specific question types and render the appropriate controls dynamically.
 
-From this base you can derive two new classes in `TextboxQuestion` and `DropdownQuestion`
-that represent textbox and dropdown questions.
-The idea is that the form will be bound to specific question types and render the
-appropriate controls dynamically.
-
-`TextboxQuestion` supports multiple HTML5 types such as text, email, and url
-via the `type` property.
+* The `TextboxQuestion` control supports multiple HTML5 types such as text, email, and url
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply aio/content/guide/dynamic-form.md:84]: It also feels like the `control` term here might be a bit confusing. The `TextboxQuestion` and `DropdownQuestion` classes are more like data models that would later be used to create actual form controls. In the current version of this section, the `control` term does not refer to the `TextboxQuestion` and `DropdownQuestion` classes (which makes sense to me).

File: aio/content/guide/dynamic-form.md [hunk 4]
```diff
@@ -50,79 +59,58 @@ Bootstrap the `AppModule` in `main.ts`.
...
 
+The dynamic form uses a service to provide a specific set of controls to be collected into a group and added to the form.
+This form group consumes the metadata from the question model and allows you to specify default values and validation rules.
 
-Next is `QuestionControlService`, a simple service for transforming the questions to a `FormGroup`.
-In a nutshell, the form group consumes the metadata from the question model and
-allows you to specify default values and validation rules.
-
+The example defines a `QuestionControlService` that collects a set of individual question controls into a `FormGroup`.
+The form group consumes the metadata from the question model and allows you to specify default values and validation rules.
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:100]: In this section the `control` term also refers to the instances of question classes when in the current version the `metadata from the question model` is used (which is technically correct).

File: aio/content/guide/dynamic-form.md [hunk 5]
```diff
@@ -50,79 +59,58 @@ Bootstrap the `AppModule` in `main.ts`.
...
-
+## Compose dynamic form contents
 
+The dynamic form itself is represented by a container component, `DynamicFormComponent`, whose contents are provided by a question component, `DynamicFormQuestionComponent`.
 
-It presents a list of questions, each bound to a `<app-question>` component element.
-The `<app-question>` tag matches the `DynamicFormQuestionComponent`,
-the component responsible for rendering the details of each _individual_
-question based on values in the data-bound question object.
+The question component provides a set of questions to include in the form using the `<app-question>' tag. This component is responsible for rendering the details of each individual question based on values in the data-bound question object.
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:110]: I find the current version a bit more easier to digest, i.e.:

[there is a list of questions] -> [each question is represented via `<app-question>`] -> [`<app-question>` matches the `DynamicFormQuestionComponent`] -> [`DynamicFormQuestionComponent` represents an individual question]

May be we can rephrase first sentence to be similar to the current state? The second sentence seems fine.

File: aio/content/guide/dynamic-form.md [hunk 6]
```diff
@@ -136,70 +124,87 @@ question based on values in the data-bound question object.
 
 </code-tabs>
 
+The `DynamicFormQuestionComponent` component can present any type of question in your model.
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:127]: I'd propose to update this sentence to something like `The goal of the DynamicFormQuestionComponent is to present question types defined in your model.`.

File: aio/content/guide/dynamic-form.md [hunk 7]
```diff
@@ -136,70 +124,87 @@ question based on values in the data-bound question object.
...
+You only have two types of questions at this point but you can imagine many more.
+The `ngSwitch` statement in the template determines which type of question to display.
+The switch uses the directives `formControlName` and `formGroup` that are defined in `ReactiveFormsModule`.
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:130]:
```
The switch uses directives with the `formControlName` and `formGroup` selectors. Both directives are defined in `ReactiveFormsModule`.
```
(and may be also include links to these directives?)

File: aio/content/guide/dynamic-form.md [hunk 8]
```diff
@@ -136,70 +124,87 @@ question based on values in the data-bound question object.
...
+{@a questionnaire-data}
 
-Notice this component can present any type of question in your model.
-You only have two types of questions at this point but you can imagine many more.
-The `ngSwitch` determines which type of question to display.
+### Supply data
 
-In both components you're relying on Angular's **formGroup** to connect the template HTML to the
-underlying control objects, populated from the question model with display and validation rules.
+`DynamicFormComponent` expects a list of questions in the form of an array bound to `@Input() questions`.
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:136]: We refer to the `DynamicFormComponent` here, but it was not defined/discussed yet. May be we can rephrase that to say that in this section we prepare a service that would take care of supplying data. For the purposes of this guide we'll use hard coded set of questions, but in a real world app this service might fetch data from the backend/storage.

File: aio/content/guide/dynamic-form.md [hunk 9]
```diff
@@ -136,70 +124,87 @@ question based on values in the data-bound question object.
...
- The set of questions you've defined for the job application is returned from the `QuestionService`.
- In a real app you'd retrieve these questions from storage.
+## Create a dynamic form template
 
- The key point is that you control the hero job application questions
- entirely through the objects returned from `QuestionService`.
- Questionnaire maintenance is a simple matter of adding, updating,
- and removing objects from the `questions` array.
+`DynamicFormComponent` is the entry point and the main container for the form.
+The selector, `<app-dynamic-form>`, places the form in the template.
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:153]: I'd propose updating it to say something like:
```
The `DynamicFormComponent` component is the entry point and the main container for the form, which is represented using the `<app-dynamic-form>` in a template.
```

**Review by AndrewKushnir — CHANGES_REQUESTED** [review]
Thanks for the PR @jbogarthyde, it provides a better structure to this guide 👍 

I've added a few comments, the main one is around the use of the `control` term. I've set the status of my review to "requested changes", so we can revisit/update the `control` term usage.

FYI, I have not tested the code in this guide as I don't see that anything code-wise has changed.

File: aio/content/guide/dynamic-form.md (hunk 2 above)
Comment (jbogarthyde) [reply aio/content/guide/dynamic-form.md:20]: The whole section is rewritten to follow the step structure - please check again.

File: aio/content/guide/dynamic-form.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbogarthyde) [reply aio/content/guide/dynamic-form.md:84]: Tried to clarify, please check again

File: aio/content/guide/dynamic-form.md (hunk 4 above)
Comment (jbogarthyde) [reply aio/content/guide/dynamic-form.md:100]: Another try. Please check again

File: aio/content/guide/dynamic-form.md [hunk 10]
```diff
@@ -17,14 +17,24 @@ you can create the new forms on the fly without changing the application code.
...
 The tutorial walks you through the following steps.
 
-1. Set up the data model behind the questionaire, defining some heroes and their attributes.
-2. Define a single, generic question component, using a `formGroup` with different control types and input validation.
-3. Compose a set of question components into a dynamic-form container component.
-4. Collect specific question details from the data model to pass into a dynamic form instance.
-5. Display the dynamic form using the predefined question data, with a Submit button that is only enabled when all user input is valid.
+1. Set up a project.
+2. Create the form object model.
+3. Compose the form contents with some sample data.
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:22]: I'd like to propose a bit more simplified version of the list and have the following items there:

1. Enable Forms feature for a project.
2. Establish data model to represent form controls.
3. Populate the model with sample data.
4. Develop a component to create form controls dynamically.

File: aio/content/guide/dynamic-form.md [hunk 11]
```diff
@@ -79,38 +83,37 @@ The following `QuestionBase` is a base class for a set of controls that can repr
...
-
-* The `TextboxQuestion` control supports multiple HTML5 types such as text, email, and url
-through the `type` property.
+When you create the form template in the next step, you will instantiate these specific question types in order to render the appropriate controls dynamically.
 
+* The `TextboxQuestion` control type presents a question and allows users to enter input.
 
 <code-example path="dynamic-form/src/app/question-textbox.ts" header="src/app/question-textbox.ts"></code-example>
 
+   The `TextboxQuestion` control type will be represented in a form template using an <input> element.
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:92]: It looks like `<input>` needs additional escaping, since it's actually rendered as an HTML element on a page, [see here](https://pr36465-90d2a2e.ngbuilds.io/guide/dynamic-form).

File: aio/content/guide/dynamic-form.md [hunk 12]
```diff
@@ -79,38 +83,37 @@ The following `QuestionBase` is a base class for a set of controls that can repr
...
+   The `type` attribute of the element will be defined based on the `type` field specified in the `options` argument (for example `text`, `email`, `url`).
+
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:94]: We may need to add a space in between these lines, since the next line is very close to the line above which makes is look like it's a part of the same logical unit. Do we have some examples of this layout (unordered list with example sections) in other places to see if we can use some code from there?

File: aio/content/guide/dynamic-form.md [hunk 13]
```diff
@@ -79,38 +83,37 @@ The following `QuestionBase` is a base class for a set of controls that can repr
...
-The dynamic form itself is represented by a container component, `DynamicFormComponent`, whose contents are provided by a question component, `DynamicFormQuestionComponent`.
-
-The question component provides a set of questions to include in the form using the `<app-question>' tag. This component is responsible for rendering the details of each individual question based on values in the data-bound question object.
+The dynamic form itself will be represented by a container component, which you will add in a later step.
+Each question is represented in the form component's template by an `<app-question>` tag, which matches an instance of `DynamicFormQuestionComponent`.
 
-The dynamic form relies on a `formGroup` to connect the template HTML to the
-underlying control objects. The `DynamicFormQuestionComponent` populate the form group with controls defined in the question model, specifying display and validation rules.
+The `DynamicFormQuestionComponent` is responsible for rendering the details of an individual question based on values in the data-bound question object.
+The form relies on a `formGroup` to connect the template HTML to the
```
Comment (AndrewKushnir) [inline aio/content/guide/dynamic-form.md:114]: Let's may be also add a link to a corresponding directive here:
```
The form relies on a [`formGroup`](api/forms/FormGroupDirective "FormGroupDirective") directive ...
```

**Review by AndrewKushnir — CHANGES_REQUESTED** [review]
Thanks for additional changes @jbogarthyde! 👍 

I've added a few comments and it looks like we need to fix the layout a bit (the most important one is with the `input` element), so I set the status of this review to "Request changes".

File: aio/content/guide/dynamic-form.md [hunk 14]
```diff
@@ -17,14 +17,24 @@ you can create the new forms on the fly without changing the application code.
 
 The tutorial walks you through the following steps.
 
-1. Set up the data model behind the questionaire, defining some heroes and their attributes.
-2. Define a single, generic question component, using a `formGroup` with different control types and input validation.
-3. Compose a set of question components into a dynamic-form container component.
-4. Collect specific question details from the data model to pass into a dynamic form instance.
-5. Display the dynamic form using the predefined question data, with a Submit button that is only enabled when all user input is valid.
+1. Set up a project.
```
[the remark this answers is not in the document]
Comment (jbogarthyde) [reply aio/content/guide/dynamic-form.md:20]: Good idea 👍

File: aio/content/guide/dynamic-form.md (hunk 11 above)
Comment (jbogarthyde) [reply aio/content/guide/dynamic-form.md:92]: oops

File: aio/content/guide/dynamic-form.md (hunk 10 above)
Comment (jbogarthyde) [reply aio/content/guide/dynamic-form.md:22]: This is just what I was hoping for - it helps to know the subject matter!

**Comment by jbogarthyde** [comment]
Done! All good ideas 👍

**Review by AndrewKushnir — APPROVED** [review]
Thanks for additional changes @jbogarthyde 👍 The changes LGTM.
